cmd/control_loop/controller/controller.go

```diff
@@ -44,59 +57,50 @@ func ProcessManifest(manifest *Manifest) ([]string, error) {
 	return actions, nil
 }
 
-func populateDependencyMaps(
+func generateDependencyMap(
 	ctx context.Context,
 	client connection.Client,
 	resourcePattern string,
-	dependencies []Dependency,
-	dependencyMaps *[]map[string]*ResourceCollection) error {
-	for _, d := range dependencies {
-		sourceMap := make(map[string]*ResourceCollection)
-		// Extend the source pattern if it contains $resource.api like pattern
-		sourcePattern, err := ParseSourcePattern(resourcePattern, d.Source)
-		if err != nil {
-			return err
-		}
+	sourcePattern string,
+	sourceFilter string) (map[string]*ResourceCollection, error) {
+
+	sourceMap := make(map[string]*ResourceCollection)
+
+	// Extend the source pattern if it contains $resource.api like pattern
+	extSourcePattern, err := ExtendSourcePattern(resourcePattern, sourcePattern)
+	if err != nil {
+		return nil, err
+	}
 
-		// Fetch resources using the sourcePattern
-		sourceList, err :=  list.ListResources( ctx, client, sourcePattern, d.Filter)
+	// Fetch resources using the extSourcePattern
+	sourceList, err :=  list.ListResources(ctx, client, extSourcePattern, sourceFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, source := range sourceList {
+		group, err := ExtractGroup(sourcePattern, source)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		// Extract the attribute which will be used
-		// to group the contents of the sourceList into a map
-		groupFuncName := ParseGroupFunc(d.Source)
-		
-		// Build source map
-		for _, source := range sourceList {
-			// Group all the resources in one group by default
-			groupBy := ""
-
-			if len(groupFuncName) > 0 {
-				groupFunc := reflect.ValueOf(source).MethodByName(groupFuncName)
-				groupBy = groupFunc.Call([]reflect.Value{})[0].String()
-			}
-
-			// Update the map with timestamp
-			source_ts := source.GetUpdateTimestamp()
-			if collection, ok := sourceMap[groupBy]; !ok {
-				sourceMap[groupBy] = &ResourceCollection {
-					maxUpdateTime: source_ts,
-					resourceList: &[]resources.Resource{},
-				}
-			} else if (*collection).maxUpdateTime.Before(source_ts) {
-					(*collection).maxUpdateTime = source_ts
+		// Update the map with timestamp
+		source_ts := source.GetUpdateTimestamp()
```

Comment:
Do we even need to keep track of maxUpdateTime as we fill up the collection in the map? Adding a function to the ResourceCollection type that can compute the maximum time seems like it would be easier and more difficult to introduce a bug to later on.

Reply:
I like this idea of keeping track of timestamp in the struct itself. I will consider this in a separate PR.